### PR TITLE
Packaging DeviceConfiguration and SimpleConfig Applications

### PR DESCRIPTION
**Pull request by kb1lqc**
This PR updates the Device Configuration and SimpleConfig applications to be imported into the python package for `faraday`. It may not be implemented the cleanest (see --fxxxxx options) or the fact that simpleconfig isn't integrated with deviceconfiguration but it is pretty straightforward and we can improve in later updates. The distribution files have been uploaded to TestPyPi to allow pip installation from the test repository. Alternatively you can install this git branch in editable mode with `pip install -e .` from the main folder.

@kb1lqd @el-iso  @jbemenheiser  @reillyeon

# Change Log
- Moved deviceconfiguration into `--faraday-deviceconfiguration`
- Moved simpleconfig into `--faraday-simpleconfig`
- Moved configuration files `deviceconfiguration.sample.ini` and `faraday_config.sample.ini` into etc/faraday
- Updated setup.cfg to reflect new version + new etc/files with commands for these two programs
- Changed `deviceconfiguration.ini` [DEVICES] options for use with one radio at a time. This simplified the code and really we only want to program one radio at a time right now. Future updates can expand this.
- Added [PROXY] section to `deviceconfiguration.sample.ini`
- Added appropriate loggers for Device Configuration and SimpleConfig programs to `loggingConfig.ini`
- Updated **ALL programs** loggers in `loggingConfig.ini` to INFO level
- Added command line options for `faraday-deviceconfiguration` and `faraday-simpleconfig`
- Added functionality to simpleconfig to read only and not program

# Command Line Options
## Device Configuration
```
$ faraday-deviceconfiguration --help
usage: faraday-deviceconfiguration [-h] [--init-config]
                                   [--init-faraday-config] [--start]
                                   [--proxycallsign PROXYCALLSIGN]
                                   [--proxynodeid PROXYNODEID]
                                   [--faradayconfig] [--callsign CALLSIGN]
                                   [--nodeid NODEID] [--redledtxon]
                                   [--redledtxoff] [--unitconfigured]
                                   [--unitunconfigured] [--gpiop3on GPIOP3ON]
                                   [--gpiop3off GPIOP3OFF] [--gpiop3clear]
                                   [--gpiop4on GPIOP4ON]
                                   [--gpiop4off GPIOP4OFF] [--gpiop4clear]
                                   [--gpiop5on GPIOP5ON]
                                   [--gpiop5off GPIOP5OFF] [--gpiop5clear]
                                   [--gpiop5 GPIOP5]
                                   [--bootfrequency BOOTFREQUENCY]
                                   [--bootrfpower BOOTRFPOWER]
                                   [--latitude LATITUDE]
                                   [--longitude LONGITUDE]
                                   [--latitudedir LATITUDEDIR]
                                   [--longitudedir LONGITUDEDIR]
                                   [--altitude ALTITUDE] [--gpsbooton]
                                   [--gpsbootoff] [--gpsenabled]
                                   [--gpsdisabled] [--uarttelemetryenabled]
                                   [--uarttelemetrydisabled]
                                   [--rftelemetryenabled]
                                   [--rftelemetrydisabled]
                                   [--uartinterval UARTINTERVAL]
                                   [--rfinterval RFINTERVAL]

Device Configuration application provides a Flask server to program Faraday
radios via an API

optional arguments:
  -h, --help            show this help message and exit
  --init-config         Initialize Device Configuration configuration file
  --init-faraday-config
                        Initialize Faraday configuration file
  --start               Start device configuration server
  --proxycallsign PROXYCALLSIGN
                        Set Proxy Faraday callsign to connect to and program
  --proxynodeid PROXYNODEID
                        Set Proxy Faraday nodeid to connect to and program
  --faradayconfig       Display Faraday configuration file contents
  --callsign CALLSIGN   Set Faraday radio callsign
  --nodeid NODEID       Set Faraday radio nodeid
  --redledtxon          Set Faraday radio RED LED during RF transmissions ON
  --redledtxoff         Set Faraday radio RED LED during RF transmissions OFF
  --unitconfigured      Set Faraday radio configured bit ON
  --unitunconfigured    Set Faraday radio configured bit OFF
  --gpiop3on GPIOP3ON   Set Faraday radio GPIO port 3 bits on, specify bit to
                        turn ON
  --gpiop3off GPIOP3OFF
                        Set Faraday radio GPIO port 3 bits on, specify bit to
                        turn OFF
  --gpiop3clear         Reset Faraday radio GPIO port 3 bits to OFF
  --gpiop4on GPIOP4ON   Set Faraday radio GPIO port 4 bits on, specify bit to
                        turn ON
  --gpiop4off GPIOP4OFF
                        Set Faraday radio GPIO port 4 bits on, specify bit to
                        turn OFF
  --gpiop4clear         Reset Faraday radio GPIO port 4 bits to OFF
  --gpiop5on GPIOP5ON   Set Faraday radio GPIO port 5 bits on, specify bit to
                        turn ON
  --gpiop5off GPIOP5OFF
                        Set Faraday radio GPIO port 5 bits on, specify bit to
                        turn OFF
  --gpiop5clear         Reset Faraday radio GPIO port 5 bits to OFF
  --gpiop5 GPIOP5       Set Faraday radio fgpio_p5
  --bootfrequency BOOTFREQUENCY
                        Set Faraday radio boot frequency
  --bootrfpower BOOTRFPOWER
                        Set Faraday radio boot RF power
  --latitude LATITUDE   Set Faraday radio default latitude. Format "ddmm.mmmm"
  --longitude LONGITUDE
                        Set Faraday radio default longitude. Format
                        "dddmm.mmmm"
  --latitudedir LATITUDEDIR
                        Set Faraday radio default latitude direction (N/S)
  --longitudedir LONGITUDEDIR
                        Set Faraday radio default longitude direction (E/W)
  --altitude ALTITUDE   Set Faraday radio default altitude in meters. Maximum
                        of 17999.99 Meters
  --gpsbooton           Set Faraday radio GPS boot power ON
  --gpsbootoff          Set Faraday radio GPS boot power OFF
  --gpsenabled          Set Faraday radio GPS use ON
  --gpsdisabled         Set Faraday radio GPS use OFF
  --uarttelemetryenabled
                        Set Faraday radio UART Telemetry ON
  --uarttelemetrydisabled
                        Set Faraday radio UART Telemetry OFF
  --rftelemetryenabled  Set Faraday radio RF Telemetry ON
  --rftelemetrydisabled
                        Set Faraday radio RF Telemetry OFF
  --uartinterval UARTINTERVAL
                        Set Faraday radio UART telemetry interval in seconds
  --rfinterval RFINTERVAL
                        Set Faraday radio RF telemetry interval in seconds

```

## SimpleConfig
```
$ faraday-simpleconfig --help
usage: faraday-simpleconfig [-h] [--read]

SimpleConfig sends a request to faraday-deviceconfiguration to initiate a POST
or GET command resulting in programming a Faraday radio and/or reading its
FLASH memory configuration

optional arguments:
  -h, --help  show this help message and exit
  --read      Read FLASH configuration only, do not program
```

# Installing from TestPyPi
As of this PR creation the latest package on TestPyPi is [faraday-0.0.1014.dev1068](https://testpypi.python.org/pypi/faraday/0.0.1014.dev1068)

Install with `pip install --no-cache-dir --extra-index-url https://testpypi.python.org/pypi faraday`

Ensure you are installing this correct version per this PR

### CODE REVIEW

**Comment by reillyeon** [comment]
Skimming through the help info it seems unnecessary to prefix all the options with "f".

**Comment by kb1lqc** [comment]
@reillyeon I actually prefix the options which change actual faraday programmed memory with an `--f`. Therefore as it currently stands `--callsign` changes the device configuration program callsign setting while `--fcallsign` changes the intended callsign to be programmed to faraday. Any other suggestions?

**Comment by kb1lqd** [comment]
@kb1lqc Why not just use the terms `--proxycallsign` to change local connection callsign and just `--callsign` for the device configuration update callsign? I agree with @reillyeon that prepending `f` before each command is unnecessary.

**Comment by kb1lqc** [comment]
OK I'll update shortly @kb1lqd  @reillyeon

**Comment by kb1lqd** [comment]
@kb1lqc How does `faraday-deviceconfiguration --fconfigboot` work? I can't seem to get it to do anything. Also this should not simply be the only way to edit the configbootbitmask. We should have a way to set the bitmask manually to a value (byte) and/or also support the "RF RED LED" bit 1 as well.

File: faraday/deviceconfiguration.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+logger.debug('faraday_config.ini PATH: ' + faradayConfigPath)
+
+# Load Device Configuration Configuration from deviceconfiguration.ini file
+deviceConfig = ConfigParser.RawConfigParser()
+
+# Command line input
+parser = argparse.ArgumentParser(description='Device Configuration application provides a Flask server to program Faraday radios via an API')
+parser.add_argument('--init-config', dest='init', action='store_true', help='Initialize Device Configuration configuration file')
+parser.add_argument('--init-faraday-config', dest='initfaraday', action='store_true', help='Initialize Faraday configuration file')
+parser.add_argument('--callsign', help='Set Proxy Faraday callsign to connect to and program')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: This should be `--proxycallsign` or similar and allow the removal of `f` prepending the other arguments.

File: faraday/deviceconfiguration.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+parser.add_argument('--nodeid', type=int, help='Set Proxy Faraday nodeid to connect to and program')
+parser.add_argument('--faradayconfig', action='store_true', help='Display Faraday configuration file contents')
+
+# Faraday Configuration
+parser.add_argument('--fcallsign', help='Set Faraday radio callsign')
+parser.add_argument('--fnodeid', type=int, help='Set Faraday radio nodeid')
+parser.add_argument('--fconfigboot', action='store_false', help='Set Faraday radio config boot bit OFF')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: 1) This doesn't seem to work... Also how do you toggle this back ON? (This is BIT 0)
2) Need to add for BIT 1 the toggling of the RED LED action on RF TX

If both bits are high then config boot bitmask = 3 (00000011)

File: faraday/deviceconfiguration.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+parser.add_argument('--fgpiop5', type=int, help='Set Faraday radio fgpio_p5')
+parser.add_argument('--fbootfrequency', type=float, help='Set Faraday radio boot frequency')
+parser.add_argument('--fbootrfpower', type=int, help='Set Faraday radio boot RF power')
+parser.add_argument('--flatitude', type=float, help='Set Faraday radio default latitude. Format \"ddmm.mmmm\"')
+parser.add_argument('--flongitude', type=float, help='Set Faraday radio default longitude. Format \"dddmm.mmmm\"')
+parser.add_argument('--flatitudedir', help='Set Faraday radio default latitude direction (N/S)')
+parser.add_argument('--flongitudedir', help='Set Faraday radio default longitude direction (E/W)')
+parser.add_argument('--faltitude', type=float, help='Set Faraday radio default altitude in meters. Maximum of 17999.99 Meters')
+# Purposely do not allow editing of GPS altitude units
+parser.add_argument('--fgpsboot', action='store_false', help='Set Faraday radio GPS boot power OFF')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: How do you toggle back on? I suggest either allowing argument options `0` and '1' or creating an `ON` and `OFF` function arguments.

File: faraday/deviceconfiguration.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+parser.add_argument('--fgps', action='store_true', help='Set Faraday radio GPS use ON')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: How do you toggle back on? I suggest either allowing argument options `0` and '1' or creating an `ON` and `OFF` function arguments.

File: faraday/deviceconfiguration.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+parser.add_argument('--fuarttelemetry', action='store_false', help='Set Faraday radio UART Telemetry OFF')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: How do you toggle back on? I suggest either allowing argument options 0 and '1' or creating an ON and OFF function arguments.

File: faraday/deviceconfiguration.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+parser.add_argument('--frftelemetry', action='store_true', help='Set Faraday radio RF Telemetry ON')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py]: How do you toggle back off? I suggest either allowing argument options 0 and '1' or creating an ON and OFF function arguments.

**Review by kb1lqd — CHANGES_REQUESTED** [review]

**Comment by kb1lqc** [comment]
@kb1lqd maybe I misunderstood the value. You can just turn it ON/OFF. Not placing `--configboot` in the command will default to setting the INI file to a value of 1 and by specifying `--configboot` you change it to a zero. This is explained in the `--help` option.

File: faraday/deviceconfiguration.py (hunk 1 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: Changed in https://github.com/FaradayRF/Faraday-Software/pull/219/commits/1556fe743c340aa67beec7e3e4849fb19021df27

File: faraday/deviceconfiguration.py (hunk 2 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: @kb1lqd 

1. It defaults to 1 and if you specify `--configboot` the INI file is changed to a zero.
2. Looks like I implemented it wrong haha....

File: faraday/deviceconfiguration.py (hunk 3 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: Defaults to ON. I should update the help description to state this.

File: faraday/deviceconfiguration.py (hunk 4 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: Defaults to OFF, I should add more to the help statement.

File: faraday/deviceconfiguration.py (hunk 5 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: Defaults to ON, will add statement to --help description

File: faraday/deviceconfiguration.py (hunk 6 above)
Comment (kb1lqc) [reply faraday/deviceconfiguration.py]: Defaults to OFF, will add help information about this.

**Comment by kb1lqc** [comment]
@kb1lqd just responded to your questions. Overall I basically made any ON/OFF option in _faraday.ini_ default to what we have in the _faraday.sample.ini_ configuration.

I do think I completely miss-implemented the `--configboot` option. Should we converse more about how to implement each bit now or in a later PR and remove the option for the time being?

**Comment by kb1lqd** [comment]
@kb1lqc Yeah this is not a fun way to do it for several reason:
* Editing multiple units with multiple settings (or making a mistake and wanting a different option) means re initializing the whole file.
* Defaulting to a standard setting is fine but any bit toggling NEED's both ON and OFF due to the above

Also, the bitmask bytes are literally bitmask bytes so all bits in use need to be editable. This is not acceptable for a later PR since things like the RED LED toggle are very useful and should be supported off the bat as they currently are.

**Comment by kb1lqc** [comment]
@kb1lqd what are the bitmask values? what does each bit represent?

**Comment by kb1lqd** [comment]
[Flash_Info_D_Memory_Allocations.xlsx](https://github.com/FaradayRF/Faraday-Software/files/1118082/Flash_Info_D_Memory_Allocations.xlsx)


We should bring a lot of this non-public facing documentation to public as it should. I think this fell through the cracks until we started needing to reference it again...

File: faraday/deviceconfiguration.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+parser.add_argument('--nodeid', type=int, help='Set Faraday radio nodeid')
+parser.add_argument('--configboot', action='store_false', help='Set Faraday radio config boot bit OFF')
+parser.add_argument('--gpiop3', type=int, help='Set Faraday radio fgpio_p3')
+parser.add_argument('--gpiop4', type=int, help='Set Faraday radio fgpio_p4')
+parser.add_argument('--gpiop5', type=int, help='Set Faraday radio fgpio_p5')
+parser.add_argument('--bootfrequency', type=float, help='Set Faraday radio boot frequency')
+parser.add_argument('--bootrfpower', type=int, help='Set Faraday radio boot RF power')
+parser.add_argument('--latitude', type=float, help='Set Faraday radio default latitude. Format \"ddmm.mmmm\"')
+parser.add_argument('--longitude', type=float, help='Set Faraday radio default longitude. Format \"dddmm.mmmm\"')
+parser.add_argument('--latitudedir', help='Set Faraday radio default latitude direction (N/S)')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py:91]: This allows other characters besides N/S. Filter?

File: faraday/deviceconfiguration.py [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+parser.add_argument('--longitudedir', help='Set Faraday radio default longitude direction (E/W)')
```
Comment (kb1lqd) [inline faraday/deviceconfiguration.py:92]: This allows other characters besides E/W. Filter?

**Review by kb1lqd — CHANGES_REQUESTED** [review]

**Comment by kb1lqc** [comment]
One thing I noticed @kb1lqd 

image

We treat the GPS bitmask and telemetry boot bitmask different. I am forced in python to create the integer value from bits for the configboot value but you accept discrete values in the configuration file for items like GPS boot and stuff. Should we be keeping these the same?

**Comment by kb1lqd** [comment]
We should standardize this but not in this pr. More of a nice too have than
lacking functionality

**Comment by kb1lqc** [comment]
@kb1lqd I implemented the GPIO functionality as well as boot bit mask functionality. It's a tad brute forced in some regards so I'll make an issue to specify that we should clean this up and add functions to pragmatically handle the GPIO bitmask creation. Would be a good beginner task to keep open that doesn't affect functionality.

**Review by kb1lqd — APPROVED** [review]
Besides a few potential updates to clarify/improve in the future the code works as expected.

**Pull request merged**